file/src/main/scala/akka/stream/alpakka/file/scaladsl/LogRotatorSink.scala

```diff
@@ -0,0 +1,119 @@
+            fileOpenOptions: Set[StandardOpenOption] = Set(StandardOpenOption.APPEND, StandardOpenOption.CREATE))
+    : LogRotatorSink =
+    new LogRotatorSink(functionGeneratorFunction, fileOpenOptions)
+}
+
+class LogRotatorSink(functionGeneratorFunction: () => ByteString => Option[Path],
+                     fileOpenOptions: Set[StandardOpenOption])
+    extends GraphStageWithMaterializedValue[SinkShape[ByteString], Future[Done]] {
+
+  val in = Inlet[ByteString]("FRotator.in")
+  override val shape = SinkShape.of(in)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[Done]) = {
+    val promise = Promise[Done]()
+    val logic = new GraphStageLogic(shape) {
+      val pathGeneratorFunction: ByteString => Option[Path] = functionGeneratorFunction()
+      var sourceOut: SubSourceOutlet[ByteString] = _
+      var fileSinkCompleted: Future[IOResult] = _
+
+      //init stage where we are waiting for the first path
+      setHandler(in,
+        new InHandler {
+        override def onPush(): Unit = {
+          val data = grab(in)
+          val pathO = pathGeneratorFunction(data)
+          if (pathO.isEmpty) {
+            pull(in)
+          } else {
+            switchPath(pathO.head, data)
+          }
+        }
+
+        override def onUpstreamFinish(): Unit =
+          completeStage()
+
+        override def onUpstreamFailure(ex: Throwable): Unit =
+          failStage(ex)
+      })
+
+      //we must pull the first element cos we are a sink
+      override def preStart(): Unit = {
+        super.preStart()
+        pull(in)
+      }
+
+      //we recreate the tail of the stream, and emit the data for the next req
+      def switchPath(path: Path, data: ByteString): Unit = {
+        if (sourceOut ne null) {
+          sourceOut.complete()
+        }
+        sourceOut = new SubSourceOutlet[ByteString]("FRotatorSource")
+        sourceOut.setHandler(new OutHandler {
+          override def onPull(): Unit = {
+            sourceOut.push(data)
+            switchToNormalMode()
+          }
+        })
+        fileSinkCompleted = Source
+          .fromGraph(sourceOut.source)
+          .runWith(FileIO.toPath(path, fileOpenOptions))(interpreter.subFusingMaterializer)
+      }
+
+      //we change path if needed or push the grabbed data
+      def switchToNormalMode(): Unit = {
+        setHandler(in,
+          new InHandler {
+          override def onPush(): Unit = {
+            val data = grab(in)
+            val pathO = pathGeneratorFunction(data)
+            if (pathO.isEmpty) {
+              sourceOut.push(data)
+            } else {
+              switchPath(pathO.head, data)
+            }
+          }
+
+          override def onUpstreamFinish(): Unit = {
+            sourceOut.complete()
+            fileSinkCompleted onComplete {
+              case Success(x) => promise.trySuccess(Done)
+              case Failure(ex) => promise.trySuccess(Done)
```

Comment:
Ah, good point, hmm, yeah, I think that it should fail the stage and the promise even if this is just closing downstream. I think it would be wrong to let that go unnoticed. Note that it will require an async-handler, it's not safe touching internal stage stuff from future callbacks.